src/Compiler/Driver/ReuseTcResults/CachingDriver.fs

```diff
@@ -0,0 +1,139 @@
+
+        tcDataFile.WriteAllLines lines
+
+    let readPrevTcData () =
+        if FileSystem.FileExistsShim tcDataFilePath then
+            use tcDataFile = FileSystem.OpenFileForReadShim tcDataFilePath
+
+            let cmdLine = ResizeArray<string>()
+            let graph = ResizeArray<string>()
+            let refs = ResizeArray<string>()
+
+            let mutable currentHeader = ""
+
+            tcDataFile.ReadLines()
+            |> Seq.iter (fun line ->
+                match line with
+                | line when line.StartsWith "BEGIN" -> currentHeader <- line.Split ' ' |> Array.last
+                | line ->
+                    match currentHeader with
+                    | CmdLineHeader -> cmdLine.Add line
+                    | GraphHeader -> graph.Add line
+                    | ReferencesHeader -> refs.Add line
+                    | _ -> invalidOp "broken tc cache")
+
+            Some
+                {
+                    CmdLine = cmdLine.ToArray()
+                    Graph = graph.ToArray()
+                    References = refs.ToArray()
+                }
+
+        else
+            None
+
+    let getThisCompilationCmdLine args = args
+
+    // maybe split into two things?
+    let getThisCompilationGraph inputs =
+        let sourceFiles =
+            inputs
+            |> Seq.toArray
+            |> Array.mapi (fun idx (input: ParsedInput) ->
+                {
+                    Idx = idx
+                    FileName = input.FileName
+                    ParsedInput = input
+                })
+
+        let filePairs = FilePairMap sourceFiles
+        let graph, _ = DependencyResolution.mkGraph filePairs sourceFiles
+
+        let list = List<string>()
+
+        for KeyValue(idx, _) in graph do
+            let fileName = sourceFiles[idx].FileName
+            let lastWriteTime = FileSystem.GetLastWriteTimeShim fileName
+            list.Add $"%i{idx},{fileName},{lastWriteTime.ToOADate()}"
+
+        for KeyValue(idx, deps) in graph do
+            for depIdx in deps do
+                list.Add $"%i{idx} --> %i{depIdx}"
+
+        list.ToArray()
+
+    let getThisCompilationReferences =
+        List.map (fun (r: AssemblyReference) -> r.Text)
+        >> List.map (fun fileName -> fileName, FileSystem.GetLastWriteTimeShim fileName)
+        >> List.map (fun (fileName, lastWriteTime) -> $"{fileName},{lastWriteTime.ToOADate()}")
+        >> List.toArray
+
+    member _.TryReuseTcResults inputs =
+        let prevTcDataOpt = readPrevTcData ()
+
+        let thisTcData =
+            {
+                CmdLine = getThisCompilationCmdLine tcConfig.cmdLineArgs
+                Graph = getThisCompilationGraph inputs
+                References = getThisCompilationReferences tcConfig.referencedDLLs
+            }
+
+        if prevTcDataOpt.IsSome then
+            use _ = Activity.start Activity.Events.reuseTcResultsCachePresent []
+
+            if prevTcDataOpt.Value = thisTcData then
```

Comment:
Which equivalency will/shall be used?

Reply:
I think the default one is sufficient for now.

Reply:
Will it be structural in this case, or referential?

Reply:
Well, `TcData` here is a record and records do structural equality, so that applies here...

Reply:
Oh, got it, thought it's a custom type.